### PR TITLE
Added ability to detect if a conn is idled and also removed some defe…

### PR DESCRIPTION
**Pull request by oxtoacart**
…rs for performance

Related to getlantern/lantern-internal#2067

### CODE REVIEW

File: idletiming_conn.go [hunk 1]
```diff
@@ -94,8 +113,12 @@ func (c *IdleTimingConn) TimesOutIn() time.Duration {
 // Read implements the method from io.Reader
 func (c *IdleTimingConn) Read(b []byte) (int, error) {
 	c.closeMutex.RLock()
-	defer c.closeMutex.RUnlock()
```
Comment (oxtoacart) [inline idletiming_conn.go]: This was a drive-by. Defers are actually quite expensive, so doing them on the critical path isn't great. This addresses that.

File: idletiming_conn.go [hunk 2]
```diff
@@ -53,6 +69,7 @@ func Conn(conn net.Conn, idleTimeout time.Duration, onIdle func()) *IdleTimingCo
 				continue
 			case <-timer.C:
 				c.Close()
+				atomic.StoreInt64(&c.idled, 1)
```
Comment (joesis) [inline idletiming_conn.go]: `c.Close()` sets `c.closed` to true and closes the wrapped connection. All r/w afterward returns `ErrIdled` (except the first `Read()` which returns `EOF`?), which is already a good indicator. In which case we need to actively check `IsIdled()`?

File: idletiming_conn.go (hunk 1 above)
Comment (joesis) [reply idletiming_conn.go]: Nice catch!

Comment (oxtoacart) [reply idletiming_conn.go]: It was thanks to your benchmarking!

File: idletiming_conn.go (hunk 2 above)
Comment (oxtoacart) [reply idletiming_conn.go]: We've been seeing a lot of EOF errors in our probes and I'm curious if any of them are caused by a connection having been idled. That's what I need IsIdled() for.

Comment (merlinran) [reply idletiming_conn.go]: I see. LGTM. Merging

**Comment by joesis** [comment]
LGTM. Merging

**Comment by oxtoacart** [comment]
Thanks @joesis!
